discord/ext/commands/cooldowns.py

```diff
@@ -287,8 +302,16 @@ def copy(self) -> DynamicCooldownMapping:
     def valid(self) -> bool:
         return True
 
-    def create_bucket(self, message: Message) -> Cooldown:
-        return self._factory(message)
+    async def create_bucket(self, ctx: Context | ApplicationContext) -> Cooldown:
+        from ...ext.commands import Context
+
+        if isinstance(ctx, Context):
+            result = self._factory(ctx.message)
+        else:
+            result = self._factory(ctx)
+        if inspect.isawaitable(result):
+            return await result
+        return result
```

Comment:
This all can be simplified with:
```suggestion
        return await utils.maybe_coroutine(self._factory, ctx)
```
Requires importing ``from discord import utils``